### PR TITLE
Remove previous implementation of LRU cache in favour of MemoryCache

### PR DESCRIPTION
**Pull request by sav007**
- Replace LRUCache with new MemoryCache
- Remove external dependency to `com.nytimes.android:cache`
- Clean up `NormalizedCache` from inheritance hell

### CODE REVIEW

File: apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/NormalizedCache.kt [hunk 1]
```diff
@@ -134,9 +79,7 @@ abstract class NormalizedCache {
     leafCache.nextCache = cache
   }
 
-  open fun dump(): Map<@JvmSuppressWildcards KClass<*>, Map<String, Record>> {
-    return mapOf(this::class to emptyMap())
-  }
+  abstract fun dump(): Map<@JvmSuppressWildcards KClass<*>, Map<String, Record>>
```
Comment (martinbonnin) [inline apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/NormalizedCache.kt:82]: 👍

File: apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/NormalizedCache.kt [hunk 2]
```diff
@@ -107,16 +62,6 @@ abstract class NormalizedCache {
...
 
-  /**
-   * Remove cached record by the key
-   *
-   * @param cacheKey of record to be removed
-   * @return `true` if record with such key was successfully removed, `false` otherwise
-   */
-  fun remove(cacheKey: CacheKey): Boolean {
-    return remove(cacheKey, false)
-  }
```
Comment (martinbonnin) [inline apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/NormalizedCache.kt:118]: 👍

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/MemoryCache.kt [hunk 3]
```diff
@@ -50,40 +52,53 @@ class MemoryCache(
...
     }
+
+    return changedKeys + nextCache?.merge(record, cacheHeaders).orEmpty()
+  }
+
+  override fun merge(records: Collection<Record>, cacheHeaders: CacheHeaders): Set<String> {
+    if (cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE)) {
+      return emptySet()
+    }
+    return records.flatMap { record -> merge(record, cacheHeaders) }.toSet()
```
Comment (martinbonnin) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/MemoryCache.kt:95]: I was going to write above that maybe we should keep the common `merge(collection)` in the parent class but given the implementation is so small, it doesn't make sense indeed. 👍

**Comment by martinbonnin** [comment]
How much of a burden would it be to keep the guava cache as a separate artifact? That could be useful if we want to do benchmarks at some point.

**Comment by sav007** [comment]
Sure I was thinking if we want to keep it, I guess the only reason is to keep it for benchmarking.
I will revert removing old LRUCache so we can benchmark it later, but keep MemoryCache as default one.

**Review by martinbonnin — APPROVED** [review]

**Pull request merged**